### PR TITLE
Resolution: Allow 3:2 pulldown and next integer and doubled upper integer

### PR DESCRIPTION
**Pull request by Memphiz**
The new whitelisted refresh rate approach changed behaviour in a way that results in a regression for macOS users. macOS has no fractional rates. So those users depend on automated switching to 24hz for 23.xx FPS content and tend to use sync to display in addition.

This change restores that refreshrate matching behaviour.

23.xx matches to 24hz
29.xx matches to 30hz or 60hz

## Motivation and Context
Regression reported in https://forum.kodi.tv/showthread.php?tid=332653

## How Has This Been Tested?

User gb160 confirmed the fix

Thx @fritsch for the code.

## Types of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by lrusak** [comment]
Seems fine. My only suggestion is to add logging in between the selections like there is above. This makes it easier to know what's going on and what's being tried/matched

File: xbmc/windowing/Resolution.cpp [hunk 1]
```diff
@@ -168,6 +168,71 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, bool is
     }
   }
 
+  // Allow 3:2 pull down
```
Comment (FernetMenta) [inline xbmc/windowing/Resolution.cpp]: limit selection explicitly to the examples you gave.

File: xbmc/windowing/Resolution.cpp [hunk 2]
```diff
@@ -168,6 +168,71 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, bool is
...
+        (info.dwFlags & D3DPRESENTFLAG_MODEMASK) == (desktop_info.dwFlags & D3DPRESENTFLAG_MODEMASK) &&
+        MathUtils::FloatEquals(info.fRefreshRate, fps * 2.5f, 0.01f))
+    {
+      CLog::Log(LOGDEBUG, "Matched 3:2 pulldown whitelisted Resolution %s (%d)", info.strMode.c_str(), i);
+      resolution = i;
+      return;
+    }
+  }
+
+  // Allow next upper integer refreshrate - some computers, like the MAC don't have fractional refreshrates - those refreshrate
```
Comment (FernetMenta) [inline xbmc/windowing/Resolution.cpp]: must not be taken if passthough audio is active

File: xbmc/windowing/Resolution.cpp [hunk 3]
```diff
@@ -168,6 +168,71 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, bool is
...
+        (info.dwFlags & D3DPRESENTFLAG_MODEMASK) == (desktop_info.dwFlags & D3DPRESENTFLAG_MODEMASK) &&
+        MathUtils::FloatEquals(info.fRefreshRate, fps , 0.1f))
+    {
+      CLog::Log(LOGDEBUG, "Matched upper sync refreshrate from whitelisted Resolution %s (%d)", info.strMode.c_str(), i);
+      resolution = i;
+      return;
+    }
+  }
+
+  // And finally to help people to get 29.97i to 60 hz
```
Comment (FernetMenta) [inline xbmc/windowing/Resolution.cpp]: explicitly limit to this case

**Comment by FernetMenta** [comment]
see my comments. exceptions must be limited to known cases. we don't want any undesired magic elsewhere.

File: xbmc/windowing/Resolution.cpp [hunk 4]
```diff
@@ -168,6 +168,71 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, bool is
...
+  // 23.976 * 2.5 ~ 59.94
+  for (const auto &mode : indexList)
+  {
+    auto i = CDisplaySettings::GetInstance().GetResFromString(mode.asString());
+    const RESOLUTION_INFO info = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo(i);
+
+    const RESOLUTION_INFO desktop_info = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo(CDisplaySettings::GetInstance().GetCurrentResolution());
+
+    if (info.iScreenWidth == desktop_info.iWidth &&
+        info.iScreen == desktop_info.iScreen &&
```
Comment (Rechi) [inline xbmc/windowing/Resolution.cpp]: `iScreen` is removed in master

**Comment by MartijnKaijser** [comment]
@Memphiz if you have some time to finish this :)

**Comment by Memphiz** [comment]
Done ;) - this is not an accepted solution. Bringing the passthrough dependency into the equation is neither.

**Pull request closed without merging**